packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,239 @@
+                            <pl-icon icon="shield-check"></pl-icon>
+                            <div class="stretch ellipsis">${$l("Password Audit")}</div>
+                        </div>
+                    </pl-button>
+                </header>
+                <div class="layout padded">
+                    <div class="vertical spacing">
+                        <pl-scroller class="stretch">
+                            <div class="counts">
+                                ${this._renderSection(_reusedPasswords, $l("Reused Passwords"), {
+                                    auditReused: "true",
+                                })}
+                                ${this._renderSection(_weakPasswords, $l("Weak Passwords"), { auditWeak: "true" })}
+                                ${this._renderSection(_compromisedPasswords, $l("Compromised Passwords"), {
+                                    auditCompromised: "true",
+                                })}
+                            </div>
+                        </pl-scroller>
+                    </div>
+                </div>
+            </div>
+        `;
+    }
+
+    async audit() {
+        const reusedPasswords: ListItem[] = [];
+        const weakPasswords: ListItem[] = [];
+        const compromisedPasswords: ListItem[] = [];
+
+        const { vaults } = this.state;
+
+        const firstMatchPerPasswordHash: {
+            [passwordHash: string]: { item: VaultItem; vault: Vault; passwordField: Field };
+        } = {};
+        const reusedPasswordItemIds: Set<string> = new Set();
+        const weakPasswordItemIds: Set<string> = new Set();
+        // const compromisedPasswordItemIds: Set<string> = new Set();
+
+        for (const vault of vaults) {
+            for (const item of vault.items) {
+                const passwordFields = item.fields.filter((field) => field.type === FieldType.Password);
+
+                if (passwordFields.length === 0) {
+                    continue;
+                }
+
+                let isReused = false;
+                let isWeak = false;
+                // let isCompromised = false;
+
+                for (const passwordField of passwordFields) {
+                    const passwordHash = await sha1(passwordField.value);
+
+                    // Perform reused audit
+                    if (Object.keys(firstMatchPerPasswordHash).includes(passwordHash)) {
+                        isReused = true;
+
+                        // Don't add the same item twice to the list, if there are more than one reused password fields in it
+                        if (!reusedPasswordItemIds.has(item.id)) {
+                            reusedPasswords.push({ item, vault });
+                            reusedPasswordItemIds.add(item.id);
+                        }
+
+                        // TODO: Save audit match boolean in field
+                        // passwordField.auditResult = { ...(passwordField.auditResult || {}), isReused };
```

Comment:
@MaKleSoft does something like this (field schema-wise) look alright to you? I'm not sure we need this to be stored in the DB, but at least it should be kept in memory.

Reply:
I think it would be better to store audit results on the item, instead of the field, as some results may affect multiple fields (e.g. the HIBP result would affect a username/password pair) and some may not affect any particular fields at all (stale/unused/expired items). How about something like this :

```ts
// @padloc/core/src/item.ts

export enum AuditResultType {
    WeakPassword = "weak_password",
    ReusedPassword = "reused_password",
    LeakedCredentials = "leaked_credentials",
}

export interface AuditResult {
    type: AuditResultType,
    /** Indexes of affected fields */
    fields: number[],
}

export class VaultItem {
    ...

    auditResults: AuditResult = [];
}
```

This gives us enough flexibility to add more audit types later without doing any major refactoring.

Reply:
With that model in mind, I think I'd also use a different schema for the filter:
 
```ts
export interface ItemsFilter {
    vault?: VaultID;
    tag?: Tag;
    favorites?: boolean;
    attachments?: boolean;
    recent?: boolean;
    host?: boolean;
    audit?: AuditType;
}
```

Reply:
@MaKleSoft Thanks for sharing your thoughts here. Maybe I'm misunderstanding something, but HIBP doesn't give you a way to check a combination of username/password. A given account/email was either leaked or not, and a given password was compromised or not.

I like your `AuditResult[]` suggestion better than mine, but I still think we should have store it on the field level (so we can show some visual indicators) as well as the item level (so we can filter). If you feel strongly (since you know the db, app, and use cases much better than me) that we should only store it in the item, we can always run an audit when an item is open and it has an audit result, to see which fields "fail".

Reply:
>  Maybe I'm misunderstanding something, but HIBP doesn't give you a way to check a combination of username/password.

Nope, I'm the one who misunderstood 😅 I guess I misremembered how the HIBP api works. Nevermind! I guess in that case we can use

```ts
fieldIndex?: number;
```
instead of 
```ts
fields: number[];
```

Since I can't think of any other audit types that might affect multiple fields.

Reply:
> I like your AuditResult[] suggestion better than mine, but I still think we should have store it on the field level (so we can show some visual indicators) as well as the item level (so we can filter). If you feel strongly (since you know the db, app, and use cases much better than me) that we should only store it in the item, we can always run an audit when an item is open and it has an audit result, to see which fields "fail".

I don't see the point in storing the result in both, since we can always do:

```ts
fieldResults = item.auditResults.filter(res => res.fieldIndex === fieldIndex);
```

No need to re-run the audit.

Reply:
Ok. I can get behind that! It does mean we need to re-run the audit when an item changes the order of fields, _but_ we'd have to run it anyway because the password fields might change on an update, so I think this works well.